### PR TITLE
Add support for Vauno EN8822C

### PR DESCRIPTION
**Pull request by gumbald**
Some work to be done on CRC and humidity needs moving to 7 bits. Have commented out battery while we're unsure where it is (copied from Esperanza EWS).

### CODE REVIEW

**Comment by zuckschwerdt** [comment]
Humidity should be `b[3] >> 1` (top 7 bits from byte 3), right?
And checksum was nibble addition IIRC?
```
int chk = ((b[4] & 0x0f) << 2) | (b[5] & 0x03);
if (add_nibbles(b, 4) != chk) {
   // checksum error
}
```

**Comment by gumbald** [comment]
Made updates as above - is there an easy way for me to check that it works with the output? I'm also a bit confused at the width/limit values, different message captures had similar but different values!

**Comment by zuckschwerdt** [comment]
- The `uint8_t b[5];` needs to be `uint8_t b[6];`
- `"mic", "Integrity", DATA_STRING, "CRC",` needs to be `"mic", "Integrity", DATA_STRING, "CHECKSUM",`
- You need to change `r_device esperanza_ews` to `r_device vauno_en8822c`
- Approximate timings are ok. E.g. here 2000, 4000, 5000, 9500

**Comment by zuckschwerdt** [comment]
- I'd say 4200, 9100 is too narrow. Round up by 10-20%, e.g. 5000, 9500.
- Fix the trailing whitespace error (line 29)
- There is no bit offset, right? Remove `uint8_t b[6];` and replace `bitbuffer_extract_bytes();` by `uint8_t *b = bitbuffer->bb[row];`
- The row selection seems wrong (L50 to 66), a bitbuffer looks like below. You'll want `int row = bitbuffer_find_repeated_prefix(bitbuffer, 4, 42)`, see https://github.com/merbanan/rtl_433/blob/master/include/bitbuffer.h#L139-L142 and other devices how it's used to get and check `row`.

```
bitbuffer:: Number of rows: 13
[00] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[01] { 0}                   :
[02] { 0}                   :
[03] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[04] { 1} 00                : 0
[05] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[06] { 1} 00                : 0
[07] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[08] { 1} 00                : 0
[09] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[10] { 1} 00                : 0
[11] {42} 85 10 f9 74 0c 40 : 10000101 00010000 11111001 01110100 00001100 01
[12] { 0}                   :
```

**Comment by gumbald** [comment]
Thanks - these are all making sense! Had tried to work out what the whitespace error was myself but couldn't track that down.

**Comment by zuckschwerdt** [comment]
> I've set the minimum as 4 as per your suggestion - could this vary? I think I've always seen 6

The idea is to require (only) a majority: `6/2+1` -- this allows for bad reception of some packets

I guess you can now remove `if (bitbuffer->num_rows != 13)` as that might vary with bad reception and just the (minimum 4 identical rows with valid checksum look solid enough.

Line 59 `uint8_t *b = bitbuffer->bb[row]` needs a `;` at the end ;)

**Comment by gumbald** [comment]
Much more Python than C recently!

**Comment by zuckschwerdt** [comment]
> Much more Python than C recently!

;) Your identation was spot on!

**Comment by zuckschwerdt** [comment]
Line 59 needs to be `data_t *data = data_make(` -- should be all good then.

**Comment by zuckschwerdt** [comment]
Looks good (ignore the "symbol errors", that's another decoder).

**Comment by merbanan** [comment]
LGTM

File: src/devices/vauno_en8822c.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    Byte:      0        1        2        3        4
+    Nibble:    1   2    3   4    5   6    7   8    9   10   11
+    Type:      IIIIIIII ??CCTTTT TTTTTTTT HHHHHHH? FFFFXXXX XX
+
+- I: Random device ID
+- C: Channel (1-3)
+- T: Temperature (Little-endian)
+- H: Humidity (Little-endian)
+- F: Flags (unknown)
+- X: Checksum
```
Comment (merbanan) [inline src/devices/vauno_en8822c.c:29]: Maybe elaborate that the checksum  is a 6 bit large nibble sum.

File: src/devices/vauno_en8822c.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    (at your option) any later version.
+*/
+/**
+Largely the same as Esperanza EWS, s3318p.
+@sa esperanza_ews.c s3318p.c
+
+List of known supported devices:
+- EN8822C-1
+
+Frame structure:
```
Comment (merbanan) [inline src/devices/vauno_en8822c.c:18]: Maybe add message size (in bits).

**Comment by gumbald** [comment]
By reading in temp_raw as int, will this handle the Two's complement? Or do I need to declare as signed int, or convert...?

**Comment by zuckschwerdt** [comment]
Yes, you will need to align the bits to 16 bit, then use the sign extend to go from 16 bit to int width.
```
int temp_raw   = (int16_t)(((b[2] & 0x0f) << 12) | ((b[2] & 0xf0) << 4) | ((b[1] & 0x0f) << 4)); // use sign extend
float temp_c  = (temp_raw >> 4) * 0.1f;
```

**Comment by gumbald** [comment]
Just wanted to make sure, is there anything else I need to do on this? Is the simplest way for me to test it to build myself presumably?

**Comment by loopy321** [comment]
Thanks for your efforts on this!  Unfortunately, I do not think it is correct for my EN8822C device from [amazon ](https://www.amazon.com/gp/product/B08ZCGYVHJ) labelled on the front "EN8822 Model T21".

I tried this pull and ran into some problems where I think the alignment is off.  See attached analysis files:

This one kind of works, but the temp is not correct.  Actual temp/humidity = approximately 75F / 55%  (sorry, do not have the display) vs. reported 177.6C:
[g005_433.92M_250k.cu8.analysis.txt](https://github.com/merbanan/rtl_433/files/9924566/g005_433.92M_250k.cu8.analysis.txt)

Placed it in the freezer at 8F.  This one is shifted somehow and does not decode/report:
[g116_433.92M_250k.cu8.analysis.txt](https://github.com/merbanan/rtl_433/files/9924572/g116_433.92M_250k.cu8.analysis.txt)

**Comment by gumbald** [comment]
Could there be a different version, what if yours is converted into F rather than C?

**Comment by loopy321** [comment]
Not sure where the problem lies, but the first file has:
`[00] {42} 6d 00 f6 66 0d 00 : 01101101 00000000 11110110 01100110 00001101 00`
with the output of decoder as:
`Channel   : 1            Temperature: 177.6 C      Humidity  : 12 %          Integrity : CHECKSUM`
The second file has:
`[00] {42} 6d 0f 78 5e 01 00 : 01101101 00001111 01111000 01011110 00000001 00`

I am a noob at this, but following the ~/docs/Analysis.md using BitBench I think the structure might be more like:

```
ID:109 CH:0 TEMP_C:  246 HUM:052 CRC:00
ID:109 CH:0 TEMP_C: -136 HUM:004 CRC:00
```

https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&f=ID%3Ad%20CH%3Ah%20TEMP_C%3A12s%20xxxxxxxxxx%20HUM%3Ad%20CRC%3A8h%20%7C%208h%2016h%2016h%20&cw=4

That gives me a consistent id, channel (0, actually 0+1), temp (24.6C, -13.6C), and humidity (52%, 4%) with what I would expect from the two above data points.  Not sure what is in all of those x's, but at least probably battery.

Maybe that is what you already have identified, but my compiled version gives incorrect decoding.

**Comment by loopy321** [comment]
> Not sure where the problem lies, but the first file has: `[00] {42} 6d 00 f6 66 0d 00 : 01101101 00000000 11110110 01100110 00001101 00` with the output of decoder as: `Channel : 1 Temperature: 177.6 C Humidity : 12 % Integrity : CHECKSUM` The second file has: `[00] {42} 6d 0f 78 5e 01 00 : 01101101 00001111 01111000 01011110 00000001 00`
> 
> I am a noob at this, but following the ~/docs/Analysis.md using BitBench I think the structure might be more like:
> 
> ```
> ID:109 CH:0 TEMP_C:  246 HUM:052 CRC:00
> ID:109 CH:0 TEMP_C: -136 HUM:004 CRC:00
> ```
> 
> https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&f=ID%3Ad%20CH%3Ah%20TEMP_C%3A12s%20xxxxxxxxxx%20HUM%3Ad%20CRC%3A8h%20%7C%208h%2016h%2016h%20&cw=4
> 
> That gives me a consistent id, channel (0, actually 0+1), temp (24.6C, -13.6C), and humidity (52%, 4%) with what I would expect from the two above data points. Not sure what is in all of those x's, but at least probably battery.
> 
> Maybe that is what you already have identified, but my compiled version gives incorrect decoding.

Adding more data points with a different analysis (row 4 is from the pull .c file).  
ID and Chan are definitely correct.   Is the humidity 2x the measurement? (i.e. row 4: 124/2=62% same as in .c file)

https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&c=af%200f%20a2%207c%2001%20c0&c=6d%2000%20bf%207c%2000%2000&c=9b%2001%2011%2062%2007%20c0&c=9b%2011%2011%2062%2008%2000&c=9b%2021%2011%2062%2008%2040&f=ID%3Ad%20CH%3Ah%20TEMP_C%3A12s%20HUM%3Ad%20CRC%3A8h%20%7C%208h&cw=4

**Comment by zuckschwerdt** [comment]
At a quick glance I'd say `int humidity = ((b[3] & 0xf0) >> 3);` is a typo, it should be `int humidity = ((b[3] & 0xfe) >> 1);` (or simply `int humidity = (b[3] >> 1);`)

**Comment by loopy321** [comment]
Is there a good resource/tutorial for coding the results from the [bitbench](https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&c=af%200f%20a2%207c%2001%20c0&c=6d%2000%20bf%207c%2000%2000&c=9b%2001%2011%2062%2007%20c0&c=9b%2011%2011%2062%2008%2000&c=9b%2021%2011%2062%2008%2040&f=ID%3Ad%20CH%3Ah%20TEMP_C%3A12s%20HUM%3Ad%20CRC%3A8h%20%7C%208h&cw=4)? 
i.e.:
`ID:d CH:h TEMP_C:12s HUM:d xxxxxxxx xxxxxxxx`

Trying to use as a learning exercise.
Thanks!

**Comment by zuckschwerdt** [comment]
We have the template https://github.com/merbanan/rtl_433/blob/master/src/devices/new_template.c
Not much about converting bits in there though.

**Comment by gumbald** [comment]
I'll have another go and compile tonight, unless you get there with the fix first!

**Comment by loopy321** [comment]
I got the temperature and humidity working with the following replacing [lines 64-67](https://github.com/merbanan/rtl_433/blob/5eee133ad8a465398e85ab5bf87dd031bdbb4f2e/src/devices/vauno_en8822c.c#L64-L67).  Thanks for the suggestions!
 
```
    int temp_raw   = (((b[1] & 0x0f) <<8) | (b[2] & 0xff)); // combine bits
    float temp_c  = (int16_t)(temp_raw - ((temp_raw & 0x800) <<1)) * 0.1f ; // sign bit
    int humidity  = (b[3] >> 1);
```
Seems like there remains some problems in checksum.

**Comment by loopy321** [comment]
Any idea how to get the checksum working properly?  
In [bitbench here](https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&c=af%200f%20a2%207c%2001%20c0&c=6d%2000%20bf%207c%2000%2000&c=9b%2001%2011%2062%2007%20c0&c=9b%2011%2011%2062%2008%2000&c=9b%2021%2011%2062%2008%2040&f=ID%3A8d%20CH%3A4h%20TEMP_C%3A12s%20HUM%3A7d%20x%20xxxx%20CRC%3A6d%20xxxxxx&cf=ADD&cw=4) I see that the ID, CH, TEMP_C, HUM all work out nicely with this scheme:

`ID:8d CH:4h TEMP_C:12s HUM:7d x xxxx CRC:6d xxxxxx
` 
according to the presumed structure:
```
    Byte:      0        1        2        3        4
    Nibble:    1   2    3   4    5   6    7   8    9   10   11
    Type:      IIIIIIII ??CCTTTT TTTTTTTT HHHHHHH? FFFFXXXX XX

```
```
6d	0	f6	66	0d	0		ID:109	CH:0	TEMP_C:246	HUM:051	CRC:52	=	65
6d	0f	78	5e	1	0		ID:109	CH:0	TEMP_C:-136	HUM:047	CRC:04	=	69
af	0f	a2	7c	1	c0		ID:175	CH:0	TEMP_C:-94	HUM:062	CRC:07	=	84
6d	0	bf	7c	0	0		ID:109	CH:0	TEMP_C:191	HUM:062	CRC:00	=	64
9b	1	11	62	7	c0		ID:155	CH:0	TEMP_C:273	HUM:049	CRC:31	=	50
9b	11	11	62	8	0		ID:155	CH:1	TEMP_C:273	HUM:049	CRC:32	=	40
9b	21	11	62	8	40		ID:155	CH:2	TEMP_C:273	HUM:049	CRC:33	=	45

```

But, as you can see the CRC add result does not work out.  Any idea based on experience how to fix this?
Thanks!

**Comment by loopy321** [comment]
I see [this works](https://triq.net/bitbench#c=6d%2000%20f6%2066%200d%2000&c=6d%200f%2078%205e%2001%2000&c=af%200f%20a2%207c%2001%20c0&c=6d%2000%20bf%207c%2000%2000&c=9b%2001%2011%2062%2007%20c0&c=9b%2011%2011%2062%2008%2000&c=9b%2021%2011%2062%2008%2040&f=ID%3A8d%20CH%3A4h%20TEMP_C%3A12s%20HUM%3A7d%20%3F5b%20CRC%3A6h%20xxxxxx&cf=ADD&cl=32&cw=4) now (based on your [issue](https://github.com/merbanan/rtl_433/issues/2171) - missed that) :

`ID:8d CH:4h TEMP_C:12s HUM:7d ?5b CRC:6h xxxxxx`

**Comment by zuckschwerdt** [comment]
*side note: CRC is a specific kind of checksum, here it's not a CRC but a plain sum (addition with carry).*

**Comment by loopy321** [comment]
I am finding a lot of missed decodes due to miscalculation of the checksum.  Based on a lot of samples, it looks like decoding only works where the last two bits are 00.

For example the following gets decoded properly (presumably because the last two bits are zero):

`{42} 4b 0f 6e 64 0f 00 : 01001011 00001111 01101110 01100100 00001111 00  ==> ID:075 CH:0 TEMP_C: -146 HUM:050`

While the following had `add_nibbles(b,4)=59 ` and `chk=((b[4] & 0x0f) << 2) | (b[5] & 0x03)=56` and does not get processed (possibly because of the last two bits not being zero):

`{42} 4b 0f 6d 64 0e c0 : 01001011 00001111 01101101 01100100 00001110 11 ==> ID:075 CH:0 TEMP_C: -147 HUM:050`

Anyone see how to change the calculations so that it works in both circumstances?

**Comment by zuckschwerdt** [comment]
A typo, that should really be `int chk = ((b[4] & 0x0f) << 2) | (b[5] >> 6);` i.e. the top 2 bit, not the bottom ones.

**Comment by loopy321** [comment]
Thanks for pointing that out!  All working well now!  See pull:
https://github.com/merbanan/rtl_433/pull/2231

**Comment by zuckschwerdt** [comment]
Thanks for testing and finding the bugs ;)

Do we know for sure that the `temp_raw` calc is correct? @gumbald figured that to be the nibbles (TTT… above) in reverse order ( 6 5 4 ) but you changed that to forward order ( 4 5 6 )?

**Comment by loopy321** [comment]
Happy to help (and thanks for help with learning). :)  I am sure that [these changes](https://github.com/merbanan/rtl_433/pull/2231/commits/32d97901c0a56183f48ffbf9caee9c78a42164b9#) result in the correct temperature and humidity for all data I have seen, including the original "Sample data" provided by @gumbald in the device file:

```
    [00] {42} af 0f a2 7c 01 c0 : 10101111 00001111 10100010 01111100 00000001 11
    Sensor ID	= 175 = 0xaf
    Channel		= 0
    temp		= -93 = 0x111110100010
    TemperatureC	= -9.3
    hum    = 62% = 0x0111110
``` 
returns : `TEMP_C:-9.4	HUM:062`

which you can also see in the [BitBench ](https://github.com/merbanan/rtl_433/pull/2172#issuecomment-1304933282) referenced above.  So I figure that is two independent devices giving correct results.

**Comment by stefanhra** [comment]
I have the same sensor as @loopy321 and was getting similar results when running gumbald:master. I just tried loopy321:test-vauno and am now getting stable, consistent, and sane results.

Thanks.

**Comment by gumbald** [comment]
I must have transposed something! Thanks for the work guys!

**Pull request closed without merging**